solutions/silver/usaco-990.mdx

```diff
@@ -260,6 +264,181 @@ print(ans, file=open("berries.out", "w"))
+Since the order of these elements does not matter, we can use a selection algorithm instead of sorting to reduce the time complexity by a factor of $\log N$.
+
+**Time Complexity:** $\mathcal{O}(max(B_i) \cdot N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+int main() {
+	freopen("berries.in", "r", stdin);
+	freopen("berries.out", "w", stdout);
+	ios::sync_with_stdio(false);
+	cin.tie(0);
+
+	int N, K;
+	cin >> N >> K;
+	int B[N];
+	for (int i = 0; i < N; i++) { cin >> B[i]; }
+	int m = 0;
+	for (int s = 1;; s++) {
+		int full = 0;
+		for (int n : B) { full += n / s; }
+		full = min(full, K);
+		if (full < K / 2) { break; }
+		if (full < K) {
+			nth_element(B, B + min(N, K - full) - 1, B + N,
+			            [&](int a, int b) { return a % s > b % s; });
+		}
+		int bessie = (full - K / 2) * s;
+		for (int i = 0; i < min(N, K - full); i++) { bessie += B[i] % s; }
+		m = max(m, bessie);
+	}
+
+	cout << m << endl;
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public class Berries {
+	static int s;
+	static Random rand;
+	public static void main(String[] args) throws IOException {
+		BufferedReader r =
+		    new BufferedReader(new FileReader(new File("berries.in")));
+		rand = new Random();
+		StringTokenizer st = new StringTokenizer(r.readLine());
+		int N = Integer.parseInt(st.nextToken());
+		int K = Integer.parseInt(st.nextToken());
+		int[] B = new int[N];
+		st = new StringTokenizer(r.readLine());
+		for (int i = 0; i < N; i++) { B[i] = Integer.parseInt(st.nextToken()); }
+		int m = 0;
+		for (s = 1;; s++) {
+			int full = 0;
+			for (int n : B) { full += n / s; }
+			full = Math.min(full, K);
+			if (full < K / 2) { break; }
+			if (full < K && K - full < N) { quickselect(B, K - full); }
+			int bessie = (full - K / 2) * s;
+			for (int i = 0; i < N && i < K - full; i++) { bessie += B[i] % s; }
+			m = Math.max(m, bessie);
+		}
+		
+		PrintWriter pw = new PrintWriter(new FileWriter("berries.out"));
+		pw.println(m);
+		pw.close();
+	}
+
+	static void quickselect(int[] A, int i) {
+		int l = 0, r = A.length - 1;
+		while (l < r) {
+			int j = partition(A, l, r);
+			if (j == i) {
+				return;
+			} else if (j < i) {
+				l = j + 1;
+			} else {
+				r = j - 1;
+			}
+		}
+	}
 
+	static int partition(int[] A, int l, int r) {
+		int m = l;
+		swap(A, l + rand.nextInt(r - l), r);
+		for (int i = l; i < r; i++) {
+			if (A[i] % s > A[r] % s) {
+				swap(A, i, m);
+				m++;
+			}
+		}
+		swap(A, m, r);
+		return m;
+	}
+
+	static void swap(int[] A, int i, int j) {
+		int t = A[i];
+		A[i] = A[j];
+		A[j] = t;
+	}
+}
+```
+
+</JavaSection>
+<PySection>
+
+<Warning>
+
+Due to Python's high constant factor, this actually runs slower than the previous solution.
+
+</Warning>
+
+```py
+import sys
+import random
+
+
+def quickselect(A, i, key):
```

Comment:
type annotations & docstring telling the user what this does
also maybe have a default for key, would be nice for ppl to copy this